helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java

```diff
@@ -0,0 +1,164 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import org.apache.helix.HelixConstants;
+import org.apache.helix.HelixException;
+import org.apache.helix.HelixProperty;
+import org.apache.helix.controller.dataproviders.ResourceControllerDataProvider;
+
+/**
+ * ResourceChangeDetector implements ChangeDetector. It caches resource-related metadata from
+ * Helix's main resource pipeline cache (DataProvider) and the computation results of change
+ * detection.
+ * WARNING: the methods of this class are not thread-safe.
+ */
+public class ResourceChangeDetector implements ChangeDetector {
+
+  private ResourceChangeCache _oldCache; // cache snapshot for previous pipeline run
+  private ResourceChangeCache _newCache; // cache snapshot for this pipeline run
+
+  private Map<String, ? extends HelixProperty> _oldPropertyMap;
+  private Map<String, ? extends HelixProperty> _newPropertyMap;
+
+  // The following caches the computation results
+  private Map<HelixConstants.ChangeType, Collection<String>> _changedItems = new HashMap<>();
+  private Map<HelixConstants.ChangeType, Collection<String>> _addedItems = new HashMap<>();
+  private Map<HelixConstants.ChangeType, Collection<String>> _removedItems = new HashMap<>();
+
+  /**
+   * Compare the underlying HelixProperty objects and produce a collection of names of changed
+   * properties.
+   * @return
+   */
+  private Collection<String> getChangedItems() {
+    Collection<String> changedItems = new HashSet<>();
+    _oldPropertyMap.forEach((name, property) -> {
+      if (_newPropertyMap.containsKey(name) && !property.equals(_newPropertyMap.get(name))) {
+        changedItems.add(name);
+      }
+    });
+    return changedItems;
+  }
+
+  /**
+   * Return a collection of names that are newly added.
+   * @return
+   */
+  private Collection<String> getAddedItems() {
+    Collection<String> addedItems = new HashSet<>(_newPropertyMap.keySet());
+    addedItems.removeAll(_oldPropertyMap.keySet());
+    return addedItems;
+  }
+
+  /**
+   * Return a collection of names that were removed.
+   * @return
+   */
+  private Collection<String> getRemovedItems() {
+    Collection<String> removedItems = new HashSet<>(_oldPropertyMap.keySet());
+    removedItems.removeAll(_newPropertyMap.keySet());
+    return removedItems;
+  }
+
+  /**
+   * Based on the change type given, call the right getters for two property maps and save the
+   * references as oldPropertyMap and newPropertyMap.
+   * @param changeType
+   */
+  private void determinePropertyMapByType(HelixConstants.ChangeType changeType) {
```

Comment:
This method modifies _oldPropertyMap and _newPropertyMap for temporary usage. If someone calls this in the wrong order, the result will be wrong. Even it is private, still not good. Others might be updating this class as well in the future.

I suggest separating this method into two. One for getting the old map, one for getting the new map. And instead of implicitly keep the return value in a private field, let's just return it.